components/google_sign_in/browser/google_sign_in_throttle.cc

```diff
@@ -0,0 +1,160 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/google_sign_in/browser/google_sign_in_throttle.h"
+
+#include <utility>
+#include <vector>
+
+#include "brave/components/google_sign_in/common/google_sign_in_util.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/permission_controller_delegate.h"
+#include "content/public/browser/web_contents.h"
+#include "extensions/common/url_pattern.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "third_party/blink/public/common/permissions/permission_utils.h"
+#include "third_party/blink/public/mojom/loader/resource_load_info.mojom-shared.h"
+
+#include "ui/base/window_open_disposition.h"
+#include "url/gurl.h"
+#include "url/origin.h"
+
+using blink::URLLoaderThrottle;
+
+namespace google_sign_in {
+
+GoogleSignInThrottle::GoogleSignInThrottle(
+    const content::WebContents::Getter& wc_getter,
+    const GURL initial_url,
+    scoped_refptr<HostContentSettingsMap> settings_map)
+    : wc_getter_(wc_getter),
+      initial_url_(initial_url),
+      settings_map_(settings_map) {}
+
+void OnPermissionRequestStatus(
+    content::NavigationEntry* pending_entry,
+    content::WebContents* contents,
+    const GURL& request_initiator_url,
+    scoped_refptr<HostContentSettingsMap> content_settings,
+    URLLoaderThrottle::Delegate* delegate,
+    const std::vector<blink::mojom::PermissionStatus>& permission_statuses) {
+  HandleBraveGoogleSignInPermissionStatus(
+      contents->GetBrowserContext(), request_initiator_url, content_settings,
+      permission_statuses);
+  DCHECK_EQ(1u, permission_statuses.size());
+  const auto status = permission_statuses[0];
+  // Check if current pending navigation is the one we started out with.
+  // This is done to prevent us from accessing a deleted Delegate, if
+  // the user navigated away while the prompt was still up, or closed the
+  // window
+  if (pending_entry != contents->GetController().GetPendingEntry()) {
+    return;
+  }
+  if (status == blink::mojom::PermissionStatus::GRANTED) {
+    delegate->Resume();
+  } else if (status == blink::mojom::PermissionStatus::DENIED) {
+    delegate->CancelWithError(net::ERR_BLOCKED_BY_CLIENT);
+  }
+  // In case of ASK, we need to be careful because delegate may be deleted
+  return;
+}
+
+void GetPermissionAndMaybeCreatePrompt(
+    bool* defer,
+    content::WebContents* contents,
+    const GURL& request_initiator_url,
+    scoped_refptr<HostContentSettingsMap> content_settings,
+    URLLoaderThrottle::Delegate* delegate) {
+  // Check kGoogleLoginControlType pref and cancel request if false
+  // NOTE: This means that if the kGoogleLoginControlType permission in
+  // is turned off, all requests to kGoogleAuthPattern and kFirebaseUrlPattern
+  // will be disallowed
+  PrefService* prefs =
```

Comment:
there is a lot of duplicated logic between this and the code in BraveContentBrowserClient. This could be cleaned up with a helper that takes a callback for the actual action to take